pkg/pgmodel/metrics/database/database.go

```diff
@@ -0,0 +1,180 @@
+package database
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/jackc/pgx/v4"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/pgxconn"
+)
+
+type Engine interface {
+	Run()
+	IsRunning() bool
+}
+
+type metricsEngineImpl struct {
+	conn      pgxconn.PgxConn
+	ctx       context.Context
+	isRunning atomic.Value
+	metrics   []metricQueryWrap
+}
+
+// NewEngine creates an engine that performs database metrics evaluation every evalInterval.
+// The engine runs predefined queries that returns a BIGINT as the metric value
+// which is then set as a value to Prometheus metric.
+//
+// Note: Make sure to call this only when the database is TimescaleDB. Plain Postgres
+// will cause evaluation errors.
+func NewEngine(ctx context.Context, conn pgxconn.PgxConn) *metricsEngineImpl {
+	engine := &metricsEngineImpl{
+		conn:    conn,
+		ctx:     ctx,
+		metrics: metrics,
+	}
+	engine.isRunning.Store(false)
+	return engine
+}
+
+func (e *metricsEngineImpl) register() {
+	prometheus.MustRegister(getMetrics(e.metrics)...)
+}
+
+func (e *metricsEngineImpl) unregister() {
+	m := getMetrics(e.metrics)
+	for i := range m {
+		prometheus.Unregister(m[i])
+	}
+}
+
+func getMetrics(m []metricQueryWrap) []prometheus.Collector {
+	var metrics []prometheus.Collector
+	for i := range m {
+		metrics = append(metrics, m[i].metric)
+	}
+	return metrics
+}
+
+const (
+	timeout       = time.Minute
+	evalInterval  = time.Minute * 3
+	maxRetryWait  = time.Minute * 30
+	stopThreshold = 5
+)
+
+func (e *metricsEngineImpl) Run() {
+	e.register()
+	e.isRunning.Store(true)
+	go func() {
+		timeoutCount := 0
+		up.Set(1)
+		defer func() {
+			e.isRunning.Store(false)
+			up.Set(0)
+		}()
+		wait := evalInterval
+		for {
+			if err := e.Update(); err != nil {
+				// Consider any error as timeout since we want to limit db executions if updating metrics is erroring out.
+				timeoutCount++
+			} else {
+				// This is to reset the state of engine that was earlier prepared for handling database timeout.
+				// The moment the db is happy with the timeout (when there is no longer a timeout), hence
+				// we resume the normal wait operation and reset the timeout so that previous timeouts do not affect
+				// the next iterations.
+				timeoutCount = 0
+				wait = evalInterval
+			}
+			if timeoutCount > stopThreshold {
+				// Stop the metrics engine temporarily if the batch times out beyond stopThreshold consecutively.
+				log.Warn("msg", fmt.Sprintf("Database metric evaluation taking more time than expected. Stopping evaluation temporarily for %.0f minutes", wait.Minutes()))
+				timeoutCount = 0
```

Comment:
This algorithm doesn't seem right. Once we get over `stopThreshold` we will multiply wait time by 2 and reset `timeoutCount` meaning that wait will not grow exponentially. I guess you want to remove `timeoutCount = 0` from this part.

Reply:
We want to exponentially incement only when a particular `wait` has been tried 5 times and for all the 5 consecutive times, it is timing out. If I remove the `timeoutCount = 0`, then the very next timeout will cause another exponential increase, meaning we do not give enough chance to the db to make itself comfortable with the current timeout which is not good.

Reply:
I am not sure I follow your logic. I thought we agreed that if we fail 5 times in a row we start doing exponential backoff until we succeed. With existing code once we fail >5 times we will set wait to 2X and then get out of exponential backoff meaning that wait time is not going to grow exponentially . Shouldn't we set `timeoutCount = 0` once we get success?

Reply:
resolving as this is discussed in the call. Now, we call the backoff as `getBackoff`, no exponential backoff.